### PR TITLE
Adds a lookup for type files when generating the package.json

### PR DESCRIPTION
**Pull request by orta**
# What / Why

This PR will fill the `types` and `flow` field when TS/Flow would have inferred support for a package based on their corresponding file system structure. 

## References

The long-form discussion is https://github.com/npm/rfcs/pull/126 - I'm making it a draft to give a chance for the rest of the TS team to discuss at our language design meeting on Friday.

### CODE REVIEW

File: read-json.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
+/** @type {import("fs")} */
```
Comment (orta) [inline read-json.js:1]: I can totally drop this, but it finalizes the auto-complete in the file

**Comment by orta** [comment]
We concluded in the design meeting that it needed support for the upcoming-but-kinda-here-now-ish esmodule syntax  - so I've added tests for those too. Should be good to go now 👍

**Comment by wraithgar** [comment]
@orta I re-added the RFC to the agenda so that it can be discussed at the next RFC meeting. I don't know if there's more that needed to be done here (I don't think there was but wanted to be sure), and hope to get an ok to move forward on this.

I resolved the merge conflict, but linting is going to need to be fixed.

File: read-json.js [hunk 2]
```diff
@@ -524,6 +524,33 @@ function final (file, data, log, strict, cb) {
...
 }
 
+function fillTypes (file, data, cb) {
+  var index = data.main ? data.main : 'index.js'
+  if (data.exports && typeof data.exports === 'string') index = data.exports
+  if (data.exports && data.exports['.']) index = data.exports['.']
+
+  function switchExt (file, ext) {
+    var extless = path.join(path.dirname(file), path.basename(file, path.extname(file)))
+    return './' + extless + '.' + ext
```
Comment (wraithgar) [inline read-json.js:534]: Is this right? By default things like `main` and `files` don't prepend `./`.  It seems inappropriate here.

Comment (ljharb) [reply read-json.js:534]: old things, like main/bin/files/browser, don't require the leading `./`, but new things, namely `exports`, does require it - it seems better to go with that more robust pattern and require it. (not sure if this comment is relevant here or not)

**Pull request closed without merging**